### PR TITLE
Feat: multiple row filters

### LINKED ISSUE #1009: FEAT: Multiple Filters
**Is your feature request related to a problem? Please describe.**

When filtering a table I often wish to filter by multiple columns.

**Describe the solution you'd like**

The ability to add multiple filters. Ideally this would be a button to let me add another filter that would be applied in addition to the first.

**Additional context**

This is inspired by TablePlus, you can see mutliple filters in their documentation [here](https://docs.tableplus.com/gui-tools/filter).

### PR DESCRIPTION
**Pull request by azmy60**
Closes #1009 

Allows users to add multiple filters using the filter builder.

Animation4

Animation4
(The action buttons, add and search, stay on top when scrolling down. No need to scroll all the way up to click the search button.)

Todo:
- [x] update filter builder functions to support AND OR operations
- [x] raw filter should reflect the filter builder
- [x] check parts in TableTable that still relies on the old `this.filter`
- [x] put it in a new file FilterBuilder.vue

### CODE REVIEW

**Comment by azmy60** [comment]
@rathboma Should we put the filter builder to a new file like `src/components/tableview/FilterBuilder.vue`? The `TableTable.vue` just keeps bigger and bigger after this feature. It makes it a little difficult to maintain. After the new file, it'd probably look something like this:

```tsx
<filter-builder :filters="tab.filters" @applyFilters="applyFilters" />
<div ref="tabulator" />
```

**Comment by rathboma** [comment]
> Looking forward to the feedbacks!
> 
> @rathboma Should we put the filter builder to a new file like `src/components/tableview/FilterBuilder.vue`? The `TableTable.vue` just keeps bigger and bigger after this feature. It makes it a little difficult to maintain. After the new file, it'd probably look something like this:
> 
> ```tsx
> <filter-builder :initialFilters="initialFilters" @applyFilters="applyFilters" />
> <div ref="tabulator" />
> ```

@azmy60 Yeah I think that's a good idea. I'm a fan of 'boy scout rules' - Leave the campsite (codebase) cleaner than you found it

File: apps/studio/src/common/appdb/models/OpenTab.ts [hunk 1]
```diff
@@ -71,28 +71,26 @@ export class OpenTab extends ApplicationEntity {
...
+  public getFilters(): Nullable<TableFilter[]> {
     try {
-      if (this.filters) {
-        const result: TableFilter = JSON.parse(this.filters)
-        return _.isObject(result) ? result : null
-      }
+      if (!this.filters) return null
+      const result: TableFilter | TableFilter[] = JSON.parse(this.filters)
+      if (_.isArray(result)) return result
+      if (_.isObject(result)) return [result]
```
Comment (rathboma) [inline apps/studio/src/common/appdb/models/OpenTab.ts:87]: Great work accounting for existing data!

File: apps/studio/src/lib/db/clients/sqlserver.js [hunk 2]
```diff
@@ -76,6 +77,7 @@ export default async function (server, database) {
     selectTop: (table, offset, limit, orderBy, filters, schema, selects) => selectTop(conn, table, offset, limit, orderBy, filters, schema, selects),
     selectTopStream: (db, table, orderBy, filters, chunkSize, schema) => selectTopStream(conn, db, table, orderBy, filters, chunkSize, schema),
     selectTopSql: (table, offset, limit, orderBy, filters, schema, selects) => selectTopSql(conn, table, offset, limit, orderBy, filters, schema, selects),
+    filterSql: (filters) => buildFilterString(filters),
```
Comment (rathboma) [inline apps/studio/src/lib/db/clients/sqlserver.js:80]: Did you intend to include this in the exported object?

File: apps/studio/src/components/tableview/TableTable.vue [hunk 3]
```diff
@@ -11,8 +11,24 @@
       <div class="table-filter">
         <form
           @submit.prevent="triggerFilter"
-          class="flex flex-middle"
+          :class="{ resizable: filters.length > 1 && filterMode === 'builder' }"
+          @scroll="$event.target.classList.toggle('filter-outside-viewport', $event.target.scrollTop > 31)"
+          ref="filterGroupWrapper"
         >
+          <!-- <div -->
```
Comment (rathboma) [inline apps/studio/src/components/tableview/TableTable.vue:18]: rm commented out code :-)

File: apps/studio/src/components/tableview/TableTable.vue [hunk 4]
```diff
@@ -59,79 +75,143 @@
...
-                class="btn btn-primary btn-fab"
-                type="submit"
-                title="Filter"
-              >
-                <i class="material-icons">search</i>
-              </button>
-            </div>
           </div>
         </form>
+        <div class="filter-drag-icon"><i class="material-icons">drag_handle</i></div>
```
Comment (rathboma) [inline apps/studio/src/components/tableview/TableTable.vue:214]: You know what I think it's great without the resizing. Can we just remove the resizing? I think even adding 5-6 filters there is still loads of space

File: apps/studio/src/components/tableview/TableTable.vue [hunk 5]
```diff
@@ -59,79 +75,143 @@
...
-                <i class="material-icons">code</i>
-              </button>
-            </div>
-            <div>
-              <div class="select-wrap">
-                <select
-                  name="Filter Field"
-                  class="form-control"
-                  v-model="filter.field"
+                <button
```
Comment (rathboma) [inline apps/studio/src/components/tableview/TableTable.vue:94]: These look great. Fits in really well. Good stuff

**Review by rathboma — CHANGES_REQUESTED** [review]
Looks awesome! One bug and one minor change, but otherwise all the styles and elements you added look fantastic.

- This change broke foreign key links (eg click a film_id in the film_actor table in postgres)
[Screencast from 09-19-2023 04:58:33 PM.webm](https://github.com/beekeeper-studio/beekeeper-studio/assets/279769/13823988-b5be-4d89-bbcb-3b61f1806c90)
- Lets just remove the resizer, it's totally fine and resizing might be confusing tbh.

File: apps/studio/src/lib/db/clients/sqlserver.js (hunk 2 above)
Comment (azmy60) [reply apps/studio/src/lib/db/clients/sqlserver.js:80]: Oops, that's not intended. I was just testing out if I could use the internal `buildFilterString` to put in the raw builder.

**Comment by azmy60** [comment]
Definitely overlooked the fk keys. Changed some codes on that but forgot to test it lol.

File: apps/studio/src/components/tableview/TableTable.vue (hunk 4 above)
Comment (azmy60) [reply apps/studio/src/components/tableview/TableTable.vue:214]: Sure, I'll do that!

**Review by rathboma — APPROVED** [review]
Let's go! :shipit:

**Pull request merged**